### PR TITLE
incremental weighted average

### PR DESCRIPTION
**Pull request by DhairyaSethi**
for #33

### CODE REVIEW

File: src/contracts/MathUtils.sol [hunk 1]
```diff
@@ -96,4 +96,63 @@ library MathUtils {
...
+   */
+  function subtractFromWeightedAverage(
+    uint256 currentWeightedSum,
+    uint256 currentSumWeights,
+    uint256 newValue,
+    uint256 newValueWeight
+  ) internal view returns (uint256, uint256, uint256) {
+    // newWeightedSum, newSumWeights, weightedAvg
+    if (currentSumWeights == newValueWeight) return (0, 0, 0);
+    if (currentSumWeights < newValueWeight) revert();
```
Comment (yan-man) [inline src/contracts/MathUtils.sol:154]: perhaps revert with a custom error / require?

File: src/contracts/MathUtils.sol [hunk 2]
```diff
@@ -96,4 +96,63 @@ library MathUtils {
...
+    uint256 currentSumWeights,
+    uint256 newValue,
+    uint256 newValueWeight
+  ) internal view returns (uint256, uint256, uint256) {
+    // newWeightedSum, newSumWeights, weightedAvg
+    if (currentSumWeights == 0) {
+      return (newValue * newValueWeight, newValueWeight, newValue);
+    }
+
+    uint256 newSumWeights = currentSumWeights + newValueWeight;
```
Comment (yan-man) [inline src/contracts/MathUtils.sol:125]: perhaps if newValueWeight == 0 we can short circuit and not go thru this calc, as it won't change current

File: src/contracts/MathUtilsExtended.sol [hunk 3]
```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: BUSL-1.1
```
Comment (yan-man) [inline src/contracts/MathUtilsExtended.sol]: perhaps can add tests for extended lib? or is the plan to wait until it is confirmed to be needed first

Comment (DhairyaSethi) [reply src/contracts/MathUtilsExtended.sol]: to the latter; yeah i think so, and if we were to go ahead with it i would refactor it to not use the struct (bc solidity stores that in memory & that achieves no purpose to us, much cheaper to keep stuff in the stack) so the interface would change as well. the core logic remains the same across libs, that's why i didn't add duplicate tests yet

File: src/contracts/MathUtils.sol (hunk 2 above)
Comment (DhairyaSethi) [reply src/contracts/MathUtils.sol:125]: yep! [f7ac505](https://github.com/aave/aave-v4/pull/47/commits/f7ac5059facb36107bd1f8e6df4f3b85289c73b9)

File: src/contracts/MathUtils.sol (hunk 1 above)
Comment (DhairyaSethi) [reply src/contracts/MathUtils.sol:154]: yep wanted to add but keeping the lib generic for now, we can add an error depending on the style we choose to go with for this codebase (for ex: i really like the new require(cond, customError))

Comment (yan-man) [reply src/contracts/MathUtils.sol:154]: agreed, and would prefer the customError over strings generally

**Comment by DhairyaSethi** [comment]
@yan-man let's change base branch to [feat/rebuild](https://github.com/aave/aave-v4/tree/feat/rebuild) & merge?

**Comment by yan-man** [comment]
@DhairyaSethi I think it can go directly into main, is there any reason to go into feat/rebuild first?

**Comment by DhairyaSethi** [comment]
@yan-man yea actually if we're merging #42 rn after merged your #50 rn, it can go in main itself

**Pull request merged**